management/WFCORE-1110_Custom_HTTP_Management_Headers.adoc

```diff
@@ -0,0 +1,139 @@
+:author:            Darran Lofthouse
+:email:             darran.lofthouse@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+The HTTP management endpoint of the application server already returns a pre-defined set of HTTP headers in all of the responses that are sent to the clients, however it is sometimes desirable that additional headers can be returned to satisfy the requirements of an environment the application server is installed within.  This enhancement is to add support for an end user to specify a custom set of HTTP headers that should be returned within each response.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-1110[WFCORE-1110]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFLY-12787[WFLY-12787]
+* https://issues.jboss.org/browse/EAP7-1335[EAP7-1335]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:tterem@redhat.com[Tomas Terem]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The HTTP management interface provides access to a number of contexts, our primary contexts are accessed at the following paths: -
+
+ * `/management` - The endpoint for servicing management requests over HTTP.
+ * `/console` - The endpoint to provide the admin console initial page and complete JavaScript.
+ 
+Subsystems can also dynmically register additional contexts to dynamically provide additional functionality or content over the HTTP management interface. 
+
+As we have different endpoints it may be desirable for them to return different headers, as an example some headers are only applicable to the handling of pages or interpretation of the JavaScript so should only be applied to the `/console` endpoint whilst other headers may be applicable to both.
+
+Configuration will be added to the HTTP management endpoint to allow headers to be specified, the following will be apply when applying the headers: -
+
+ * The configuration will allow an end user to specify headers to be added based on a specified path prefix.
+ * This will be a new `constant-headers` attribute on the `/core-service=management/management-interface=http-interface` resource.
+ * Within the `constant-headers` attribute it will be possible to specify a set of headers using key / value pairs against a specific path.
+ * All paths will be assumed to be prefixes starting with `/` - any paths missing a leading `/` will be handled as though the leading `/` has been specified.
+ * If a specific header is defined multiple times it will be added to the response multiple times.
+ * Configured headers will not override any headers added by the corresponding endpoint.
```

Comment:
I think we need to clarify what these two rules means for existing headers:
Based on first rule, if we add headers {A: abc, A: 123} then both will be included in a response.
Based on second rule, if /management endpoint already adds {Content-Length: 188} and we configure constant-headers {Content-Length: abc, Content-Length: 123} then the original value won't be overriden. 

I think the analysis should clearly say what happens in this case. "will not override" could mean that only existing value would be used, but also that all three would be used.

Reply:
Going to work through this one today with one of the WIPs in relation to the header handling.

A number of headers that we already set are known to be problematic, we may just need to end up identifying a set of headers we ban from being set,

Overall all our implementation can do is pro-actively set a set of headers as the request is inbound, how the internal Undertow code, management endpoints and extenstions respond to these headers is not under our direct control.